modules/apps/transfer/keeper/relay_forwarding_test.go

```diff
@@ -612,3 +613,169 @@ Test async ack is properly relayed to middle hop after forwarding transfer compl
+	suite.Require().Equal(amount, total.Amount, errorMsg)
+}
+
+func (suite *KeeperTestSuite) TestOnTimeoutPacketForwarding() {
+	pathAtoB, pathBtoC := suite.setUpForwardingPaths()
+
+	amount := sdkmath.NewInt(100)
+	coin := sdk.NewCoin(sdk.DefaultBondDenom, amount)
+	sender := suite.chainA.SenderAccounts[0].SenderAccount
+	receiver := suite.chainC.SenderAccounts[0].SenderAccount
+
+	denomA := types.NewDenom(coin.Denom)
+	denomAB := types.NewDenom(coin.Denom, types.NewTrace(pathAtoB.EndpointB.ChannelConfig.PortID, pathAtoB.EndpointB.ChannelID))
+	denomABC := types.NewDenom(coin.Denom, types.NewTrace(pathAtoB.EndpointA.ChannelConfig.PortID, pathAtoB.EndpointA.ChannelID), types.NewTrace(pathBtoC.EndpointB.ChannelConfig.PortID, pathBtoC.EndpointB.ChannelID))
+
+	originalABalance := suite.chainA.GetSimApp().BankKeeper.GetBalance(suite.chainA.GetContext(), sender.GetAddress(), coin.Denom)
+
+	hops := &types.Forwarding{
+		Hops: []types.Hop{
+			{
+				PortId:    pathBtoC.EndpointA.ChannelConfig.PortID,
+				ChannelId: pathBtoC.EndpointA.ChannelID,
+			},
+		},
+	}
+
+	transferMsg := types.NewMsgTransfer(
+		pathAtoB.EndpointA.ChannelConfig.PortID,
+		pathAtoB.EndpointA.ChannelID,
+		sdk.NewCoins(coin),
+		sender.GetAddress().String(),
+		receiver.GetAddress().String(),
+		suite.chainA.GetTimeoutHeight(),
+		0, "",
+		hops,
+	)
+
+	result, err := suite.chainA.SendMsgs(transferMsg)
+	suite.Require().NoError(err) // message committed
+
+	// parse the packet from result events and recv packet on chainB
+	packet, err := ibctesting.ParsePacketFromEvents(result.Events)
+	suite.Require().NoError(err)
+	suite.Require().NotNil(packet)
+
+	err = pathAtoB.EndpointB.UpdateClient()
+	suite.Require().NoError(err)
+
+	// Receive packet on B.
+	result, err = pathAtoB.EndpointB.RecvPacketWithResult(packet)
+	suite.Require().NoError(err)
+	suite.Require().NotNil(result)
+
+	err = pathBtoC.EndpointA.UpdateClient()
+	suite.Require().NoError(err)
+
+	err = pathBtoC.EndpointB.UpdateClient()
+	suite.Require().NoError(err)
+
+	// Make sure founds went from A to B's escrow account.
+	suite.assertAmountOnChain(suite.chainA, balance, originalABalance.Amount.Sub(amount), denomA.IBCDenom(), "Chain A should have less funds")
+	suite.assertAmountOnChain(suite.chainB, escrow, amount, denomAB.IBCDenom(), "Chain B's escrow should have coins")
+
+	address := types.GetForwardAddress(packet.DestinationPort, packet.DestinationChannel).String()
+	data := types.NewFungibleTokenPacketDataV2(
+		[]types.Token{
+			{
+				Denom:  types.NewDenom(sdk.DefaultBondDenom, types.NewTrace(pathAtoB.EndpointA.ChannelConfig.PortID, pathAtoB.EndpointA.ChannelID)),
+				Amount: "100",
+			},
+		},
+		address,
+		suite.chainC.SenderAccounts[0].SenderAccount.GetAddress().String(),
+		"", nil,
+	)
+
+	packet = channeltypes.NewPacket(
+		data.GetBytes(),
+		1,
+		pathBtoC.EndpointA.ChannelConfig.PortID,
+		pathBtoC.EndpointA.ChannelID,
+		pathBtoC.EndpointB.ChannelConfig.PortID,
+		pathBtoC.EndpointB.ChannelID,
```

Comment:
I think basically you don't need this `data` and `packet` and with the one you construct later on should be enough. You can move it up here to be able to use and that should be it.

Reply:
> These are also not correct, because this packet should be the packet sent from A to B

I'm not sure I understand this. In my intention this was the packet that times out going from B to C, so I think we need to use the path from B to C.

Why should I need to use A to B instead? (which makes the test fail)

Reply:
Similarly as explained in my comment above, you're constructing here the packet for which an acknowledgement has been written on chain B on timeout. And that packet should be the one that was sent from chain A to chain B. So the path that you should be using is `pathAtoB` instead of `pathBtoC`.

You see it also when you do down below `err = suite.chainA.GetSimApp().TransferKeeper.OnAcknowledgementPacket(suite.chainA.GetContext(), packet, data, ack)`. Here you are relaying the acknowledgement written on chain B back to the chain A. And the `packet` that you use is the one for the transfer done between A and B.

Let me know if it all makes sense!

Reply:
It should be fine now :)